src/oneprometheus/opennebula-exporter/src/opennebula_power_collector.rb

```diff
@@ -0,0 +1,261 @@
+    #   - opennebula_vmproc_power_consumption_uW
+    # --------------------------------------------------------------------------
+    DEFAULT_PROC_INTERVAL = 5
+    DEFAULT_PROC_OVERHEAD = 0.1
+    DEFAULT_PROC_TICKER = '/var/lib/one-context/get_proc_ticks'
+
+    VMPROC_POWER_METRICS = {
+        'vmproc_power_consumption_uW' => {
+            :type   => :gauge,
+            :docstr => 'VM process estimated power usage',
+            :labels => VMPROC_LABELS
+        }
+    }
+
+    VMPROC_PERCENT_METRICS = {
+        'vmproc_cpu_time_percent' => {
+            :type   => :gauge,
+            :docstr => 'VM process CPU time percent',
+            :labels => VMPROC_LABELS
+        }
+    }
+
+    class << self
+        attr_accessor :scaph_data
+    end
+
+    def get_scaph_metrics(host, port, seconds = 2)
+        hostname = String(host['NAME'])
+        Timeout.timeout(seconds) do
+            begin
+                TCPSocket.new(hostname, port).close
+                begin
+                    res = Net::HTTP.get(hostname, '/metrics', port)
+                    self.class.scaph_data[host['ID'].to_i] = res.split("\n")
+                rescue StandardError => e
+                    puts e.message
+                    puts "Error accessing to scaphandre metrics on #{hostname}"
+                end
+            rescue Errno::ECONNREFUSED, Errno::EHOSTUNREACH
+                puts "Error on #{hostname}:#{port}, is scaphandre running there?"
+            end
+        rescue Timeout::Error
+            false
+        end
+    end
+
+    def query_vm_agent(vm_id, query, host = 'localhost')
+        cmd = "virsh -c 'qemu+ssh://#{host}/system' qemu-agent-command one-#{vm_id} '#{query.to_json}'"
+        out = `#{cmd}`
+        JSON.parse(out)
+    end
+
+    def issue_command(vm_id, host, cmd, args)
+        query_processes = { :execute => 'guest-exec',
+                            :arguments => { :path => cmd, :arg => args.split,
+                                            :'capture-output' => true } }
+        begin
+            out = query_vm_agent(vm_id, query_processes, host)
+            out['return']['pid']
+        rescue StandardError
+            -1
+        end
+    end
+
+    def retrieve_stdout_command(vm_id, pid, host)
+        query_get_stdout = { 'execute'   => 'guest-exec-status',
+                             'arguments' => { 'pid' => pid } }
+        out = query_vm_agent(vm_id, query_get_stdout, host)
+
+        begin
+            Base64.decode64(out['return']['out-data'])
+        rescue StandardError
+            ''
+        end
+    end
+
+    def get_proc_consumption(vm_id, host = localhost, full_power = 0)
+        interval = DEFAULT_PROC_INTERVAL
+        overhead = DEFAULT_PROC_OVERHEAD
+        cmd = DEFAULT_PROC_TICKER
+
+        pid_cmd = issue_command(vm_id, host, cmd, "-s #{interval} -m #{vm_id} -p #{full_power}")
+    rescue StandardError
+        puts "Error executing command on VM #{vm_id}"
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Potential log injection detected. Ensure all untrusted input is properly sanitized before logging. Use parameterized logging or validate input against an allow list to prevent log injection vulnerabilities. Consider using a dedicated logging library&#39;s built-in sanitization features when available. Learn more - https://cwe.mitre.org/data/definitions/117.html